File: Console/NAND_Block_Manager.cs
using System;
using System.Collections.Generic;
using FlashMemory;
using Microsoft.VisualBasic;

public class NAND_BLOCK_IF
{
    public event PrintConsoleEventHandler PrintConsole;

    public delegate void PrintConsoleEventHandler(string msg);

    public event SetProgressEventHandler SetProgress;

    public delegate void SetProgressEventHandler(int percent);

    public event ReadPagesEventHandler ReadPages;

    public delegate void ReadPagesEventHandler(uint page_addr, ushort page_offset, uint count, FlashArea area, ref byte[] data);

    public event WritePagesEventHandler WritePages;

    public delegate void WritePagesEventHandler(uint page_addr, byte[] main, byte[] oob, FlashArea area, ref bool write_result);

    public event EraseSectorEventHandler EraseSector;

    public delegate void EraseSectorEventHandler(uint page_addr, ref bool erase_result);

    public event ReadyEventHandler Ready;

    public delegate void ReadyEventHandler(); // Checks the RD/BSY pin or register (WAS WaitForReady)

    public List<MAPPING> MAP = new List<MAPPING>();

    public uint MAPPED_PAGES { get; set; } // Number of pages available in the current map
    private long NAND_SIZE { get; set; } // Typical size of the nand (does not include extra area)
    private uint PAGE_MAIN { get; set; } // Size of the main page
    private uint PAGE_OOB { get; set; } // Size of the ext page
    private ushort PAGE_COUNT { get; set; } // Number of pages in the block
    private ushort BLOCK_COUNT { get; set; } // Number of blocks


    // Private Property BLOCK_SIZE As UInt32 'size of the block (minus oob)

    private byte[] MEMORY_AREA_ERASED; // If CopyExtendedArea is true, this will contain the other area that was erased

    public NAND_BLOCK_IF()
    {
    }

    public enum BLOCK_STATUS
    {
        Valid, // Marked valid by the manager or by the user
        Bad_Manager, // This block was marked bad because of the manager setting
        Bad_Marked, // This block was marked bad because of the user
        Bad_ByError, // This block was marked bad because of programming
        Unknown
    }

    public class MAPPING
    {
        public BLOCK_STATUS Status;
        public uint BlockIndex; // Index of the block
        public long PageAddress;  // The physical address of the first page of this block
        public long LogicalPage; // The mapped address of the first page of this block
    }

    // Called once on device init to create a map of all blocks
    public void CreateMap(long mem_size, uint main_page, uint oob_page, ushort page_count, uint block_count)
    {
        NAND_SIZE = mem_size;
        PAGE_MAIN = main_page;
        PAGE_OOB = oob_page;
        PAGE_COUNT = page_count;
        BLOCK_COUNT = (ushort)block_count;
        MAP.Clear();
        for (uint i = 0U, loopTo = (uint)(BLOCK_COUNT - 1); i <= loopTo; i++)
        {
            var block_info = new MAPPING();
            block_info.Status = BLOCK_STATUS.Valid;
            block_info.BlockIndex = i;
            block_info.PageAddress = PAGE_COUNT * (long)i;
            block_info.LogicalPage = PAGE_COUNT * (long)i;
            MAP.Add(block_info);
        }
    }

    public bool EnableBlockManager()
    {
        try
        {
            if (MainApp.MySettings.NAND_BadBlockManager == BadBlockMode.Disabled | MainApp.MySettings.NAND_Layout == NandMemLayout.Combined)
            {
                PrintConsole?.Invoke("NAND block manager disabled");
                return true;
            }
            else
            {
                PrintConsole?.Invoke("NAND memory device detected, loading valid memory map");
            }

            uint page_addr = 0U;
            for (uint i = 0U, loopTo = (uint)(BLOCK_COUNT - 1); i <= loopTo; i++)
            {
                uint LastPageAddr = (uint)(page_addr + PAGE_COUNT - 1L); // The last page of this block
                byte[] page_one = null;
                byte[] page_two = null;
                byte[] page_last = null;
                bool valid_block = true;
                int markers = (int)MainApp.MySettings.NAND_BadBlockMarkers;
                if ((markers & (int)BadBlockMarker._1stByte_FirstPage) > 0)
                {
                    if (page_one is null)
                        ReadPages?.Invoke(page_addr, (ushort)0, 6U, FlashArea.OOB, ref page_one);
                    if (!(page_one[0] == 255))
                        valid_block = false;
                }
                if ((markers & (int)BadBlockMarker._1stByte_SecondPage) > 0)
                {
                    if (page_two is null)
                        ReadPages?.Invoke((uint)((long)page_addr + 1L), (ushort)0, 6U, FlashArea.OOB, ref page_two);
                    if (!(page_two[0] == 255))
                        valid_block = false;
                }
                if ((markers & (int)BadBlockMarker._1stByte_LastPage) > 0)
                {
                    if (page_last is null)
                        ReadPages?.Invoke(LastPageAddr, (ushort)0, 6U, FlashArea.OOB, ref page_last);
                    if (!(page_last[0] == 255))
                        valid_block = false;
                }
                if ((markers & (int)BadBlockMarker._6thByte_FirstPage) > 0)
                {
                    if (page_one is null)
                        ReadPages?.Invoke(page_addr, (ushort)0, 6U, FlashArea.OOB, ref page_one);
                    if (!(page_one[5] == 255))
                        valid_block = false;
                }
                if ((markers & (int)BadBlockMarker._6thByte_SecondPage) > 0)
                {
                    if (page_two is null)
                        ReadPages?.Invoke((uint)((long)page_addr + 1L), (ushort)0, 6U, FlashArea.OOB, ref page_two);
                    if (!(page_two[5] == 255))
                        valid_block = false;
                }
                if (!valid_block)
                {
                    PrintConsole?.Invoke(string.Format("BAD NAND BLOCK at page index: 0x{0} (block index: {1})", Conversion.Hex(page_addr).PadLeft(6, '0'), i.ToString()));
                    MAP[(int)i].Status = BLOCK_STATUS.Bad_Manager;
                }
                else
                {
                    MAP[(int)i].Status = BLOCK_STATUS.Valid;
                }

                page_addr += PAGE_COUNT;
            }

            return true;
        }
        catch
        {
        }

        return false;
    }

    public void ProcessMap()
    {
        MAPPED_PAGES = 0U;
        long Logical_Page_Pointer = 0L;
        for (uint i = 0U, loopTo = (uint)(MAP.Count - 1); i <= loopTo; i++)
        {
            if (MAP[(int)i].Status == BLOCK_STATUS.Valid)
            {
                MAP[(int)i].LogicalPage = Logical_Page_Pointer;
                Logical_Page_Pointer += PAGE_COUNT;
                MAPPED_PAGES += PAGE_COUNT;
            }
            else
            {
                MAP[(int)i].LogicalPage = 0L;
            }
        }

        PrintConsole?.Invoke(string.Format("NAND memory map complete: {0} pages available for access", Strings.Format((object)MAPPED_PAGES, "#,###")));
    }
    // Returns the physical page address from the logical page address
    public uint GetPageMapping(uint page_index)
    {
        for (uint i = 0U, loopTo = (uint)(MAP.Count - 1); i <= loopTo; i++)
        {
            if (MAP[(int)i].Status == BLOCK_STATUS.Valid)
            {
                uint page_start = (uint)MAP[(int)i].LogicalPage;
                uint page_end = (uint)(page_start + PAGE_COUNT - 1L);
                if (page_index >= page_start && page_index <= page_end)
                {
                    return (uint)(MAP[(int)i].PageAddress + (page_index - page_start));
                }
            }
        }

        return 0U; // NOT FOUND
    }
    // Returns the total space of the extra data
    public uint Extra_GetSize()
    {
        uint PageCount = (uint)(NAND_SIZE / (double)PAGE_MAIN); // Total number of pages in this device
        uint PageExtraSize = PageCount * PAGE_OOB;
        return PageExtraSize;
    }

    public bool ERASEBLOCK(long page_address, FlashArea Memory_area, bool CopyOtherArea)
    {
        MEMORY_AREA_ERASED = null;
        if (CopyOtherArea && !(Memory_area == FlashArea.All))
        {
            if (Memory_area == FlashArea.Main)
            {
                ReadPages?.Invoke((uint)page_address, (ushort)0, (uint)PAGE_COUNT * PAGE_OOB, FlashArea.OOB, ref MEMORY_AREA_ERASED);
            }
            else if (Memory_area == FlashArea.OOB)
            {
                ReadPages?.Invoke((uint)page_address, (ushort)0, (uint)PAGE_COUNT * PAGE_MAIN, FlashArea.Main, ref MEMORY_AREA_ERASED);
            }
        }

        try // START BLOCK ERASE
        {
            var result = default(bool);
            EraseSector?.Invoke((uint)page_address, ref result);
            if (!result)
                return false;
            Ready?.Invoke();
        }
        catch
        {
        }  // BLOCK ERASE COMPLETE

        return true;
    }
    // This writes the data to a page
    public bool WRITEPAGE(long page_address, byte[] data_to_write, FlashArea memory_area)
    {
        try
        {
            byte[] main_data = null;
            byte[] oob_data = null;
            var WriteResult = default(bool);
            if (memory_area == FlashArea.Main)
            {
                main_data = data_to_write;
                oob_data = MEMORY_AREA_ERASED;
            }
            else if (memory_area == FlashArea.OOB)
            {
                main_data = MEMORY_AREA_ERASED;
                oob_data = data_to_write;
            }
            else if (memory_area == FlashArea.All)
            {
                main_data = data_to_write;
            }

            MEMORY_AREA_ERASED = null;
            WritePages?.Invoke((uint)page_address, main_data, oob_data, memory_area, ref WriteResult);
            if (!WriteResult)
                return false;
            if (PAGE_MAIN == 512L) // LEGACY NAND DEVICE
            {
                Utilities.Sleep(100);
            }
            else
            {
                Utilities.Sleep(10);
            }

            Ready?.Invoke();
            return WriteResult;
        }
        catch
        {
            return false;
        }
    }

    public byte[] READPAGE(long page_addr, ushort page_offset, uint count, FlashArea memory_area)
    {
        byte[] data_out = null;
        ReadPages?.Invoke((uint)page_addr, page_offset, count, memory_area, ref data_out);
        return data_out;
    }

    public bool EraseChip()
    {
        long PageAddr = 0L;
        for (int i = 0, loopTo = BLOCK_COUNT - 1; i <= loopTo; i++)
        {
            bool preserve = MainApp.MySettings.NAND_Preserve;
            if (MainApp.NAND_LayoutTool.Layout == NandMemLayout.Combined)
                preserve = false;
            bool Result = this.ERASEBLOCK(PageAddr, FlashArea.Main, preserve);
            if (!Result)
                return false;
            if (preserve && MEMORY_AREA_ERASED is object) // We need to write back the OOB
            {
                WritePages?.Invoke((uint)PageAddr, null, MEMORY_AREA_ERASED, FlashArea.OOB, ref Result);
                MEMORY_AREA_ERASED = null;
            }

            PageAddr += PAGE_COUNT;
            if (i % 10 == 0)
            {
                int Percent = (int)Math.Round(i / (double)BLOCK_COUNT * 100d);
                SetProgress?.Invoke(Percent);
            }
        }
        SetProgress?.Invoke(0);
        return true;
    }
}

File: FlashcatUSB/NAND_Block_Manager.vb
﻿Imports FlashcatUSB.FlashMemory

'Implements a NAND Flash block manager
Public Class NAND_BLOCK_IF
    Public Event PrintConsole(msg As String)
    Public Event SetProgress(percent As Integer)
    Public Event ReadPages(page_addr As UInt32, page_offset As UInt16, count As UInt32, area As FlashArea, ByRef data() As Byte)
    Public Event WritePages(page_addr As UInt32, main() As Byte, oob() As Byte, area As FlashArea, ByRef write_result As Boolean)
    Public Event EraseSector(page_addr As UInt32, ByRef erase_result As Boolean)
    Public Event Ready() 'Checks the RD/BSY pin or register (WAS WaitForReady)
    Public MAP As New List(Of MAPPING)
    Public Property MAPPED_PAGES As UInt32 'Number of pages available in the current map
    Private Property NAND_SIZE As Long 'Typical size of the nand (does not include extra area)
    Private Property PAGE_MAIN As UInt32 'Size of the main page
    Private Property PAGE_OOB As UInt32 'Size of the ext page
    Private Property PAGE_COUNT As UInt16 'Number of pages in the block
    Private Property BLOCK_COUNT As UInt16 'Number of blocks


    'Private Property BLOCK_SIZE As UInt32 'size of the block (minus oob)

    Private MEMORY_AREA_ERASED() As Byte 'If CopyExtendedArea is true, this will contain the other area that was erased


    Sub New()

    End Sub

    Public Enum BLOCK_STATUS
        Valid 'Marked valid by the manager or by the user
        Bad_Manager 'This block was marked bad because of the manager setting
        Bad_Marked 'This block was marked bad because of the user
        Bad_ByError 'This block was marked bad because of programming
        Unknown
    End Enum

    Public Class MAPPING
        Public Status As BLOCK_STATUS
        Public BlockIndex As UInt32 'Index of the block
        Public PageAddress As Long  'The physical address of the first page of this block
        Public LogicalPage As Long 'The mapped address of the first page of this block
    End Class

    'Called once on device init to create a map of all blocks
    Public Sub CreateMap(mem_size As Long, main_page As UInt32, oob_page As UInt32, page_count As UInt16, block_count As UInt32)
        Me.NAND_SIZE = mem_size
        Me.PAGE_MAIN = main_page
        Me.PAGE_OOB = oob_page
        Me.PAGE_COUNT = page_count
        Me.BLOCK_COUNT = block_count
        MAP.Clear()
        For i As UInt32 = 0 To Me.BLOCK_COUNT - 1
            Dim block_info As New MAPPING
            block_info.Status = BLOCK_STATUS.Valid
            block_info.BlockIndex = i
            block_info.PageAddress = (CLng(Me.PAGE_COUNT) * i)
            block_info.LogicalPage = (CLng(Me.PAGE_COUNT) * i)
            MAP.Add(block_info)
        Next
    End Sub

    Public Function EnableBlockManager() As Boolean
        Try
            If MySettings.NAND_BadBlockManager = FlashcatSettings.BadBlockMode.Disabled Or MySettings.NAND_Layout = FlashcatSettings.NandMemLayout.Combined Then
                RaiseEvent PrintConsole(RM.GetString("nand_block_manager_disabled"))
                Return True
            Else
                SetStatus(RM.GetString("nand_mem_device_detected"))
                RaiseEvent PrintConsole(RM.GetString("nand_mem_map_loading"))
            End If
            Dim page_addr As UInt32 = 0
            For i As UInt32 = 0 To Me.BLOCK_COUNT - 1
                Dim LastPageAddr As UInt32 = (page_addr + PAGE_COUNT - 1) 'The last page of this block
                Dim page_one() As Byte = Nothing
                Dim page_two() As Byte = Nothing
                Dim page_last() As Byte = Nothing
                Dim valid_block As Boolean = True
                Dim markers As Integer = MySettings.NAND_BadBlockMarkers
                If (markers And FlashcatSettings.BadBlockMarker._1stByte_FirstPage) > 0 Then
                    If page_one Is Nothing Then RaiseEvent ReadPages(page_addr, 0, 6, FlashArea.OOB, page_one)
                    If Not ((page_one(0)) = 255) Then valid_block = False
                End If
                If (markers And FlashcatSettings.BadBlockMarker._1stByte_SecondPage) > 0 Then
                    If page_two Is Nothing Then RaiseEvent ReadPages(page_addr + 1, 0, 6, FlashArea.OOB, page_two)
                    If Not ((page_two(0)) = 255) Then valid_block = False
                End If
                If (markers And FlashcatSettings.BadBlockMarker._1stByte_LastPage) > 0 Then
                    If page_last Is Nothing Then RaiseEvent ReadPages(LastPageAddr, 0, 6, FlashArea.OOB, page_last)
                    If Not ((page_last(0)) = 255) Then valid_block = False
                End If
                If (markers And FlashcatSettings.BadBlockMarker._6thByte_FirstPage) > 0 Then
                    If page_one Is Nothing Then RaiseEvent ReadPages(page_addr, 0, 6, FlashArea.OOB, page_one)
                    If Not ((page_one(5)) = 255) Then valid_block = False
                End If
                If (markers And FlashcatSettings.BadBlockMarker._6thByte_SecondPage) > 0 Then
                    If page_two Is Nothing Then RaiseEvent ReadPages(page_addr + 1, 0, 6, FlashArea.OOB, page_two)
                    If Not ((page_two(5)) = 255) Then valid_block = False
                End If
                If (Not valid_block) Then
                    Application.DoEvents()
                    RaiseEvent PrintConsole(String.Format(RM.GetString("mem_bad_nand_block"), Hex(page_addr).PadLeft(6, "0"), i.ToString))
                    MAP(i).Status = BLOCK_STATUS.Bad_Manager
                Else
                    MAP(i).Status = BLOCK_STATUS.Valid
                End If
                page_addr += PAGE_COUNT
            Next
            Return True
        Catch ex As Exception
        End Try
        Return False
    End Function

    Public Sub ProcessMap()
        Me.MAPPED_PAGES = 0
        Dim Logical_Page_Pointer As Long = 0
        For i As UInt32 = 0 To MAP.Count - 1
            If MAP(i).Status = BLOCK_STATUS.Valid Then
                MAP(i).LogicalPage = Logical_Page_Pointer
                Logical_Page_Pointer += Me.PAGE_COUNT
                Me.MAPPED_PAGES += Me.PAGE_COUNT
            Else
                MAP(i).LogicalPage = 0
            End If
        Next
        RaiseEvent PrintConsole(String.Format(RM.GetString("nand_mem_map_complete"), Format(MAPPED_PAGES, "#,###")))
    End Sub
    'Returns the physical page address from the logical page address
    Public Function GetPageMapping(page_index As UInt32) As UInt32
        For i As UInt32 = 0 To MAP.Count - 1
            If (MAP(i).Status = BLOCK_STATUS.Valid) Then
                Dim page_start As UInt32 = MAP(i).LogicalPage
                Dim page_end As UInt32 = (page_start + Me.PAGE_COUNT) - 1
                If page_index >= page_start AndAlso page_index <= page_end Then
                    Return MAP(i).PageAddress + (page_index - page_start)
                End If
            End If
        Next
        Return 0 'NOT FOUND
    End Function
    'Returns the total space of the extra data
    Public Function Extra_GetSize() As UInt32
        Dim PageCount As UInt32 = NAND_SIZE / PAGE_MAIN 'Total number of pages in this device
        Dim PageExtraSize As UInt32 = (PageCount * PAGE_OOB)
        Return PageExtraSize
    End Function

    Public Function ERASEBLOCK(page_address As Long, Memory_area As FlashArea, CopyOtherArea As Boolean) As Boolean
        MEMORY_AREA_ERASED = Nothing
        If CopyOtherArea AndAlso Not Memory_area = FlashArea.All Then
            If Memory_area = FlashArea.Main Then
                RaiseEvent ReadPages(page_address, 0, (PAGE_COUNT * PAGE_OOB), FlashArea.OOB, MEMORY_AREA_ERASED)
            ElseIf Memory_area = FlashArea.OOB Then
                RaiseEvent ReadPages(page_address, 0, (PAGE_COUNT * PAGE_MAIN), FlashArea.Main, MEMORY_AREA_ERASED)
            End If
        End If
        Try 'START BLOCK ERASE
            Dim result As Boolean
            RaiseEvent EraseSector(page_address, result)
            If Not result Then Return False
            RaiseEvent Ready()
        Catch ex As Exception
        End Try  'BLOCK ERASE COMPLETE
        Return True
    End Function
    'This writes the data to a page
    Public Function WRITEPAGE(page_address As Long, data_to_write() As Byte, memory_area As FlashArea) As Boolean
        Try
            Dim main_data() As Byte = Nothing
            Dim oob_data() As Byte = Nothing
            Dim WriteResult As Boolean
            If memory_area = FlashArea.Main Then
                main_data = data_to_write
                oob_data = MEMORY_AREA_ERASED
            ElseIf memory_area = FlashArea.OOB Then
                main_data = MEMORY_AREA_ERASED
                oob_data = data_to_write
            ElseIf memory_area = FlashArea.All Then
                main_data = data_to_write
            End If
            MEMORY_AREA_ERASED = Nothing
            RaiseEvent WritePages(page_address, main_data, oob_data, memory_area, WriteResult)
            If Not WriteResult Then Return False
            If (PAGE_MAIN = 512) Then 'LEGACY NAND DEVICE
                Utilities.Sleep(100)
            Else
                Utilities.Sleep(10)
            End If
            RaiseEvent Ready()
            Return WriteResult
        Catch ex As Exception
            Return False
        End Try
    End Function

    Public Function READPAGE(page_addr As Long, page_offset As UInt16, count As UInt32, memory_area As FlashArea) As Byte()
        Dim data_out() As Byte = Nothing
        RaiseEvent ReadPages(page_addr, page_offset, count, memory_area, data_out)
        Return data_out
    End Function

    Public Function EraseChip() As Boolean
        Dim PageAddr As Long = 0
        For i = 0 To Me.BLOCK_COUNT - 1
            Dim preserve As Boolean = MySettings.NAND_Preserve
            If MySettings.NAND_Layout = FlashcatSettings.NandMemLayout.Combined Then preserve = False
            Dim Result As Boolean = ERASEBLOCK(PageAddr, FlashArea.Main, preserve)
            If Not Result Then Return False
            If preserve AndAlso Me.MEMORY_AREA_ERASED IsNot Nothing Then 'We need to write back the OOB
                RaiseEvent WritePages(PageAddr, Nothing, MEMORY_AREA_ERASED, FlashArea.OOB, Result)
                MEMORY_AREA_ERASED = Nothing
            End If
            PageAddr += PAGE_COUNT
            If i Mod 10 = 0 Then
                Dim Percent As Integer = Math.Round((i / Me.BLOCK_COUNT) * 100)
                RaiseEvent SetProgress(Percent)
            End If
        Next
        RaiseEvent SetProgress(0)
        Return True
    End Function

End Class
